apps/full-site-editing/full-site-editing-plugin/full-site-editing/plugins/editor-template-classes/index.js

```diff
@@ -0,0 +1,56 @@
+/* global fullSiteEditing */
+/**
+ * External dependencies
+ */
+import classNames from 'classnames';
+import { endsWith, get, map } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { withSelect } from '@wordpress/data';
+import { registerPlugin } from '@wordpress/plugins';
+
+const EditorTemplateClasses = withSelect( select => {
+	const { getEntityRecord } = select( 'core' );
+	const { getEditedPostAttribute } = select( 'core/editor' );
+	const templatePartClasses = map( getEditedPostAttribute( 'template_part_types' ), typeId => {
+		const typeName = get(
+			getEntityRecord( 'taxonomy', 'wp_template_part_type', typeId ),
+			'name',
+			''
+		);
+		if ( endsWith( typeName, '-header' ) ) {
+			return 'site-header site-branding';
+		}
+		if ( endsWith( typeName, '-footer' ) ) {
+			return 'site-footer';
+		}
+	} );
+	return { templatePartClasses };
+} )( ( { templatePartClasses } ) => {
+	const blockListInception = setInterval( () => {
+		const blockList = document.querySelector(
+			'.editor-block-list__layout.block-editor-block-list__layout'
+		);
+
+		if ( ! blockList ) {
+			return;
+		}
+		clearInterval( blockListInception );
+
+		blockList.className = classNames(
+			'editor-block-list__layout',
+			'block-editor-block-list__layout',
+			...templatePartClasses
+		);
```

Comment:
If we go with this approach, adding classes to the internal block-list component is pretty fragile. Though I guess there's not much else wrapping the blocks otherwise. At a minimum, we'll need an E2E test as a follow up. https://github.com/WordPress/gutenberg/blob/87d5d5a19e9f57ba6414d51c19657fb6ed6b7d81/packages/block-editor/src/components/block-list/index.js#L211

As an alternative to quickly check, does locking the group here also lock down all children? https://developer.wordpress.org/block-editor/developers/block-api/block-templates/#locking

Reply:
@mmtr did something similar to lock down editing in the page content slot where we had similar issues of content being inserted outside of the container, etc. 

It think from memory it is this https://github.com/Automattic/wp-calypso/blob/master/apps/full-site-editing/full-site-editing-plugin/full-site-editing/blocks/post-content/edit.js#L71 bit that allows the parent block to be locked, but children not. 

It also involved setting the editor root to the post content block and overriding the block inserter 
https://github.com/Automattic/wp-calypso/tree/master/apps/full-site-editing/full-site-editing-plugin/full-site-editing/editor/block-inserter

Reply:
@gwwar @glendaviesnz I'm still a bit rusty and I'm not sure I understand entirely template locking.
As far as I can see, it only affects inserting, moving, and removing blocks. (@mmtr please feel free to correct me here!)
This might help us to prevent blocks outside the template part group, but not to apply the correct theme style to them.

As a wild idea, we could add a way to "link" FSE to the current theme classes.
Say, for example, adding a site option for the header CSS class:
```php
$current_theme_name = get_option( 'stylesheet' );
$fse_header_classes_option = $current_theme_name . '-fse-header-classes';
add_option( $fse_header_classes_option, [ 'site-header', 'site-branding' ] );
```

Or, we could just make it so that FSE expects that the theme uses some "standard" classes (e.g. 'site-header', 'site-footer').

Reply:
Looking at https://github.com/Automattic/themes/pull/1167 (cc @vindl)
It could be a nice way to extend this approach.

E.g. in the current proposed diff it returns something like this:
```js
{
  header:
    '<!-- wp:group {"className":"site-header site-branding"} -->
      <div class="wp-block-group site-header site-branding">
        <div class="wp-block-group__inner-container">
          <!-- wp:a8c/site-title /-->
          <!-- wp:a8c/site-description /-->
        </div>
      </div>
    <!-- /wp:group -->',
}
```
But we could make it return this instead:
```js
{
  headerClasses: [
    'site-header',
    'site-branding',
  ],
  headerContent:
    '<!-- wp:a8c/site-title /-->
     <!-- wp:a8c/site-description /-->',
}
```

Reply:
I think that's an awesome approach. It's extensible and lets the theme do style as it needs. I think the original drawback is still there - we'd still need to manually apply the classNames to the block wrapper, right? But I don't think there's any avoiding that if Gutenberg doesn't have a way to do it.

Another thing that _could_ get

Reply:
I don't mean to block, I was curious if folks had tried it. Feel free to ship, but we'll keep an 👀 on follow up. I may also ask you to write a new E2E in a bit :D 

cc @mtias @aduth @youknowriad non-urgent, but if folks happen to know if we can lock down a container like column or group's controls (doesn't display/interactive), but still allows for insertion of children and interaction with inner blocks.

Reply:
> we'd still need to manually apply the classNames to the block wrapper, right?

Do you mean for blocks inserted _after_ the initial automagical pre-generation?
Yes, unfortunately. I can't think of a way to avoid that.
Although, maybe by providing the theme classes via API, we might eventually find a way to incept the block insertion. 🤔

Reply:
@gwwar Do you mean like an "invisible" block container, where you interact with its children but never actually see it in the UI? What exactly do you want to lock from the container?

Reply:
@gwwar see https://github.com/WordPress/gutenberg/issues/7469

Reply:
@youknowriad right, the use case is we want to provide structure, but not allow folks to interact with the container UI or container block settings, because they might break what we intend. We'd still allow folks to add inner blocks and modify those. It also avoids an extra click to what the user sees as interesting (the content) vs interacting with layout structure.

Thanks @mtias for the link to https://github.com/WordPress/gutenberg/issues/7469 !